src/node_contextify.cc

```diff
@@ -1299,10 +1362,89 @@ void ContextifyContext::CompileFunction(
                            produce_cached_data,
                            std::move(new_cached_data))
           .IsNothing()) {
-    return;
+    return Object::New(env->isolate());
   }
 
-  args.GetReturnValue().Set(result);
+  return result;
+}
+
+// When compiling as CommonJS source code that contains ESM syntax, the
+// following error messages are returned:
+// - `import` statements: "Cannot use import statement outside a module"
+// - `export` statements: "Unexpected token 'export'"
+// - `import.meta` references: "Cannot use 'import.meta' outside a module"
+// Dynamic `import()` is permitted in CommonJS, so it does not error.
+// While top-level `await` is not permitted in CommonJS, it returns the same
+// error message as when `await` is used in a sync function, so we don't use it
+// as a disambiguation.
+constexpr std::array<std::string_view, 3> esm_syntax_error_messages = {
+    "Cannot use import statement outside a module",  // `import` statements
+    "Unexpected token 'export'",                     // `export` statements
+    "Cannot use 'import.meta' outside a module"};    // `import.meta` references
```

Comment:
nit: these comments don't add any value—the messages are abundantly clear.

Reply:
Sorry, I disagree; I want to make really clear which syntax each error corresponds to. Especially since `import()` isn’t part of this. It’s also not obvious why import and export have such different messages; I wanted to make clear that these two correspond with each other despite completely different phrasing.